csrc/expr_evaluator.cpp

```diff
@@ -197,6 +197,11 @@ const PolymorphicValue& ExpressionEvaluator::evaluate(ParallelType pt) {
   return null_;
 }
 
+const PolymorphicValue& ExpressionEvaluator::evaluate(const Val* value) const {
+  const auto it = known_values_.find(value);
+  return it != known_values_.end() ? it->second : null_;
+}
+
```

Comment:
This behaves a bit different from the non-const evaluate, since it is restricted to known values only so it is "shallow" like `getValue`. Could we instead consider making `getValue` const (which it should be) and further just exposing that as a public method? That way it is clear that `getValue` does a shallow but const lookup and `evaluate` does a deep but non-const one. Alternatively, we could implement this const version of `evaluate` recursively just like the non-const one, but just avoid the step where we set values in `known_values_`.

Reply:
`getValue` is very similar to this const evaluate method. Is there any particular advantage over picking one of these implementations over the other?

Reply:
I meant that the name of this new function might be confusing since I would expect that `expr_eval.evaluate(v)` would return the same result whether or not `expr_eval` is const. I don't actually think there's a need for a public version of `getValue`, although it should probably be marked const. I would probably prefer making the const `evaluate` method behave just like its non-const version but skipping the part where it records the evaluated values to `known_values_`.

Reply:
I agree with Jacob's idea for making the constant `evaluate` do real evaluation and just not update `known_values_`